File: leetcode/diameter-of-binary-tree.py
# https://leetcode.com/problems/diameter-of-binary-tree/

# Tags: Tree - Depth-First Search - Binary Tree

import timeit
from typing import Optional

from data import TreeNode, deserializeStringArrayToBinaryTree


# We need to find the longest possible path between nodes, from each sub-tree root, we check if the longest path is in the subtree
# or it would come from the parent node
#
# Time complexity O(n) - we need to visit every node once
# Space complexity O(log(n)) - At most, the call stack will be as long as the tree is deep
#
# Runtime: 70 ms, faster than 52.49% of Python3 online submissions for Diameter of Binary Tree.
# Memory Usage: 16.4 MB, less than 10.03% of Python3 online submissions for Diameter of Binary Tree.
class DFS:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.max_length = 0

        def dfs(root: Optional[TreeNode]) -> int:
            if not root:
                return 0
            # If this node is not null, return the sum of the longest path found in each sub-tree plus the node
            # traveled to get here
            left, right = dfs(root.left), dfs(root.right)

            # If the subtree with this node as root has a longer path between nodes than from the root,
            # update it
            combined = left + right
            if combined > self.max_length:
                self.max_length = combined

            # The longest path from the parent through this node will be going through the longest branch
            return 1 + max(left, right)

        dfs(root)
        return self.max_length


def test():
    root1 = deserializeStringArrayToBinaryTree("[1,2,3,4,5]")
    root2 = deserializeStringArrayToBinaryTree("[1,2]")
    executors = [DFS]
    tests = [
        [root1, 3],
        [root2, 1],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(int(float("1"))):
            for i, t in enumerate(tests):
                sol = executor()
                result = sol.diameterOfBinaryTree(t[0])
                exp = t[1]
                assert (
                    result == exp
                ), f"\033[93m» {result} <> {exp}\033[91m for test {i} using \033[1m{executor.__name__}"
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        res = "{0:20}{1:10}{2:10}".format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()
# drawTree(deserializeStringArrayToBinaryTree("[6,2,8,0,4,7,9,null,null,3,5]"))

File: leetcode/diameter-of-binary-tree.rs
// 543. Diameter of Binary Tree
// 🟢 Easy
//
// https://leetcode.com/problems/diameter-of-binary-tree/
//
// Tags: Tree - Depth-First Search - Binary Tree

use std::cell::RefCell;
use std::rc::Rc;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

struct Solution;
impl Solution {
    /// Recursive postorder with the recursive function returning two values, the length of the
    /// longest path between the subtree root and a leave, and the number of nodes in the subtree
    /// diameter, each node uses these two values in their left and right sub-tree to compute its
    /// own return values.
    ///
    /// Time complexity: O(n) - We visit each node in the tree using postorder DFS.
    /// Space complexity: O(h) - The call stack will grow to size h, which could be equal to n.
    ///
    /// Runtime 0 ms Beats 100%
    /// Memory 2.72 MB Beats 65.15%
    #[allow(dead_code)]
    pub fn diameter_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn dfs(node: Option<Rc<RefCell<TreeNode>>>) -> (i32, i32) {
            match node {
                None => (0, 0),
                Some(rc) => {
                    let l = dfs(rc.borrow().left.clone());
                    let r = dfs(rc.borrow().right.clone());
                    (1 + l.0.max(r.0), (1 + r.0 + l.0).max(r.1.max(l.1)))
                }
            }
        }
        dfs(root).1
    }
}

// Tests.
fn main() {
    let tests = [(vec![0], 0)];
    println!("\n\x1b[92m» Running {} tests...\x1b[0m", tests.len());
    // let mut success = 0;
    // for (i, t) in tests.iter().enumerate() {
    //     let res = Solution::diameter_of_binary_tree(t.0.clone());
    //     if res == t.1 {
    //         success += 1;
    //         println!("\x1b[92m✔\x1b[95m Test {} passed!\x1b[0m", i);
    //     } else {
    //         println!(
    //             "\x1b[31mx\x1b[95m Test {} failed expected: {:?} but got {}!!\x1b[0m",
    //             i, t.1, res
    //         );
    //     }
    // }
    // println!();
    // if success == tests.len() {
    //     println!("\x1b[30;42m✔ All tests passed!\x1b[0m")
    // } else if success == 0 {
    //     println!("\x1b[31mx \x1b[41;37mAll tests failed!\x1b[0m")
    // } else {
    //     println!("\x1b[31mx\x1b[95m {} tests failed!\x1b[0m", tests.len() - success)
    // }
}
